### PR TITLE
devilutionx: unstable-2019-07-28 -> 0.5.0

### PR DESCRIPTION
**Pull request by karolchmist**
###### Motivation for this change

DevilutionX 0.5.0 release:
https://github.com/diasurgical/devilutionX/releases/tag/0.5.0

Based on Devilution 0.10.0

Features
- Sound is now accurate to the original
- All in-game issues fixed
- Delete hero, inline dialogs and scrollbars are now implemented
- Screenshots now have different names
- Multiple simultaneous dialogs fixed
- All builds are now 64bit (except for Windows and Raspberry Pi)
- Memory leaks and crashes fixed
- All keys are now mapped
- UI text now has correct shadows
- Much lower CPU usage

Known issues
- Error dialogs not implemented in main UI
- The game must restart after hosting multiplayer

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @karolchmist  (myself :P )

### CODE REVIEW

File: pkgs/games/devilutionx/default.nix [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-  version = "unstable-2019-07-28";
+  version = "0.5.0";
   pname = "devilutionx";
 
   src = fetchFromGitHub {
     owner = "diasurgical";
     repo = "devilutionX";
-    rev = "b2f358874705598ec139f290b21e340c73d250f6";
-    sha256 = "0s812km118qq5pzlzvzfndvag0mp6yzvm69ykc97frdiq608zw4f";
+    rev = "88039e9a21f708f07fc0922e5beb96820b5ad607";
```
Comment (aanderse) [inline pkgs/games/devilutionx/default.nix]: `rev = version;`?

**Review by aanderse — CHANGES_REQUESTED** [review]

File: pkgs/games/devilutionx/default.nix (hunk 1 above)
Comment (karolchmist) [reply pkgs/games/devilutionx/default.nix]: I thought that a tag can be moved, a hash stays forever :)
I've put `version`, easier to maintain.

**Comment by aanderse** [comment]
@GrahamcOfBorg build devilutionx

**Comment by aanderse** [comment]
@karolchmist are you able to address the `darwin` failure?

**Comment by karolchmist** [comment]
No, I don't see why it failed... Seems like it compiles, and then fails in install phase. I don't have macOs. What can we do? Maybe remove darwin from the platform list?

**Comment by aanderse** [comment]
`cp: cannot stat 'devilutionx': No such file or directory`

Seems like a different executable is produced on `darwin`. If we had access to a `darwin` box we could simply insert an `echo` to tell us what is in that directory and go from there...

**Comment by karolchmist** [comment]
@aanderse I pushed a commit that lists the content of the output folder. On darwin, it's :
```
[100%] Linking CXX executable devilutionx.app/Contents/MacOS/devilutionx
Copying OS X content devilutionx.app/Contents/Resources/AppIcon.icns
Copying OS X content devilutionx.app/Contents/Resources/CharisSILB.ttf
[100%] Built target devilutionx
installing
total 14932
drwxr-xr-x 16 nixbld1 nixbld      544 Oct 16 14:32 .
drwxr-xr-x 27 nixbld1 nixbld      918 Oct 16 14:30 ..
-rw-r--r--  1 nixbld1 nixbld        2 Jan  1  1970 .gitignore
-rw-r--r--  1 nixbld1 nixbld    35343 Oct 16 14:30 CMakeCache.txt
drwxr-xr-x 17 nixbld1 nixbld      578 Oct 16 14:33 CMakeFiles
-r--r--r--  1 nixbld1 nixbld     3870 Oct 16 14:30 CPackConfig.cmake
-r--r--r--  1 nixbld1 nixbld     4347 Oct 16 14:30 CPackSourceConfig.cmake
-rw-r--r--  1 nixbld1 nixbld   132032 Oct 16 14:30 Makefile
-rw-r--r--  1 nixbld1 nixbld     1731 Oct 16 14:30 cmake_install.cmake
-rw-r--r--  1 nixbld1 nixbld       77 Oct 16 14:30 config.h
drwxr-xr-x  3 nixbld1 nixbld      102 Oct 16 14:30 devilutionx.app
-rw-r--r--  1 nixbld1 nixbld    18512 Oct 16 14:32 libPKWare.a
-rw-r--r--  1 nixbld1 nixbld    43848 Oct 16 14:32 libRadon.a
-rw-r--r--  1 nixbld1 nixbld   103104 Oct 16 14:32 libStormLib.a
-rw-r--r--  1 nixbld1 nixbld 14875984 Oct 16 14:32 libdevilution.a
-rw-r--r--  1 nixbld1 nixbld    42976 Oct 16 14:30 libsmacker.a
```

It seems like `devilutionx.app` is the executable file. 

Do you know what we should do in this case?

**Comment by aanderse** [comment]
@karolchmist unfortunately I've never really used `darwin` much before so even the most basic specifics of the system elude me.

ping @lilyball as someone who might be generous and share basic `darwin` knowledge with us :heart:

**Comment by aanderse** [comment]
Maybe @jsamsa wouldn't mind taking 30 seconds to explain an extremely basic `darwin` concept to us so you can finish this PR up... :thinking:

**Comment by jsamsa** [comment]
mv devilutionx.app $out/Applications

**Comment by jsamsa** [comment]
There might be a binary that could (should) be moved to $out/bin as well. I'm not familiar with this package so, not sure myself. Have a look at `pkgs/development/tools/electron/default.nix` as an example

**Comment by karolchmist** [comment]
Thanks for your help @aanderse and @jsamsa . I tried to make sense of how to build it on Darwin, but I gave up because I wasn't sure what I was doing. Even if it builds, I can't test it properly. 

If one day someone is willing to make it work, I'd be happy to help. Until then, I prefer to remove the support for Darwin. It's a niche game, and I don't think anybody cares...

To finish on a positive note: I re-enabled the hardening for the `format` flag, since it passes for the current version of the project.

**Comment by aanderse** [comment]
@karolchmist alright. Too bad, but we don't have the hardware so if someone wants `darwin` support they'll have to add themselves to the maintainers list I guess.

Can you please format your commit messages as per nixpkgs standard? As in:
- `devilutionx: unstable-2019-07-28 -> 0.5.0`
- `devilutionx: Remove darwin platform due to build errors`
- `devilutionx: Remove hardeningDisable = [ "format" ] `

Or squash all commits into:
- `devilutionx: unstable-2019-07-28 -> 0.5.0`

Thanks!

**Comment by lilyball** [comment]
@jsamsa's suggestion looks correct. The output as listed in your comment is `devilutionx.app`, which is a folder structure that contains the executable (among other things). The whole `devilutionx.app` is the GUI app and should be moved into `$out/Applications` as already recommended. I don't see anything in that output that obviously needs to be moved elsewhere either.

**Comment by lilyball** [comment]
I don't have `diabdat.mpq` but I was able to run the app on darwin and get the screen telling me it's missing (I am curious how it's supposed to work with a Nix install though; where does the file go?)

Here's the diff applied to get to this point:

```diff
diff --git a/pkgs/games/devilutionx/default.nix b/pkgs/games/devilutionx/default.nix
index 63149aff678..650d12c112b 100644
--- a/pkgs/games/devilutionx/default.nix
+++ b/pkgs/games/devilutionx/default.nix
@@ -17,10 +17,13 @@ stdenv.mkDerivation rec {
 
   installPhase = ''
     runHook preInstall
-
+  '' + (if stdenv.isDarwin then ''
+    mkdir -p $out/Applications
+    mv devilutionx.app $out/Applications
+  '' else ''
     mkdir -p $out/bin
     cp devilutionx $out/bin
-
+  '') + ''
     runHook postInstall
   '';
 
@@ -29,6 +32,6 @@ stdenv.mkDerivation rec {
     description = "Diablo build for modern operating systems";
     license = licenses.unlicense;
     maintainers = [ maintainers.karolchmist ];
-    platforms = platforms.linux ++ platforms.windows;
+    platforms = platforms.linux ++ platforms.darwin ++ platforms.windows;
   };
 }
```

**Comment by aanderse** [comment]
Thank you for your detailed explanation @lilyball! You are exceptional, as always! Generally how these open source forks/remakes of proprietary games work is that they ask you to put the game data in a predefined location like `~/.local/share/GAME`.

@karolchmist at this point @lilyball has handed us the solution... If you don't mind can you **please** keep `darwin` support?

**Comment by lilyball** [comment]
My question with diabdat.mpq is because the instructions say

> Copy diabdat.mpq from your CD, or GoG install folder, to the DevilutionX install folder; Make sure it is all lowercase.

That makes it sound like it's supposed to be sitting next to the executable, which obviously doesn't work with Nix (for any platform).

That said, after digging through all of the code it does appear that if it can't find the file next to the executable it falls back to `$HOME/.local/share/diasurgical/devilution/`, so the Nix install should work fine even on macOS. I just wish this was documented.

**Comment by karolchmist** [comment]
Hey @lilyball, thanks for you help! I pushed a version with your solution for Darwin.

How would you document the location for `diabdat.mpq`? Just in the description?

**Comment by aanderse** [comment]
`longDescription` should be fine.

**Comment by aanderse** [comment]
@GrahamcOfBorg build devilutionx

File: pkgs/games/devilutionx/default.nix [hunk 2]
```diff
@@ -1,35 +1,38 @@
...
     cp devilutionx $out/bin
+  '') + ''
 
     runHook postInstall
   '';
 
   meta = with stdenv.lib; {
     homepage = "https://github.com/diasurgical/devilutionX";
     description = "Diablo build for modern operating systems";
+    longDescription = "To make it work, it is necessary to put diabdat.mpq file in ~/.local/share/diasurgical/devilution";
```
Comment (aanderse) [inline pkgs/games/devilutionx/default.nix]:
```suggestion
    longDescription = "In order to play this game a copy of diabdat.mpq is required. Place a copy of diabdat.mpq in ~/.local/share/diasurgical/devilution before executing the game.";
```

**Comment by aanderse** [comment]
@karolchmist small suggestion on wording. Looking good. Thanks again @lilyball :tada:

**Pull request merged**